crates/rpc/rpc/src/eth/helpers/sync_listener.rs

```diff
@@ -0,0 +1,130 @@
+//! A utility Future to asynchronously wait until a node has finished syncing.
+
+use futures::Stream;
+use pin_project::pin_project;
+use reth_network_api::NetworkInfo;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Future that resolves once the node is no longer syncing.
+#[must_use = "futures do nothing unless polled"]
+#[pin_project]
+#[derive(Debug)]
+pub struct SyncListener<N, St> {
+    #[pin]
+    tick: St,
+    network_info: N,
+}
+
+impl<N, St> SyncListener<N, St> {
+    /// Create a new [`SyncListener`] using the given tick stream.
+    pub const fn new(network_info: N, tick: St) -> Self {
+        Self { tick, network_info }
+    }
+}
+
+impl<N, St> Future for SyncListener<N, St>
+where
+    N: NetworkInfo,
+    St: Stream<Item = ()> + Unpin,
```

Comment:
I think we could even relax the Item type here with an additional generic impl<N,St,Out> ?